ashura/engine/animation.h

```diff
@@ -2,25 +2,620 @@
+  } grid;
+  std::function<f32(i32 row, i32 col, i32 index, i32 total)> easing = nullptr;
+};
+
+static std::pair<i32, i32> grid_position(i32 index, i32 rows, i32 cols,
+                                         GridDirection direction)
+{
+  if (direction == GridDirection::RowMajor)
+  {
+    return {
+        index / cols,        // row
+        index % cols         // col
+    };
+  }
+  else
+  {
+    return {
+        index % rows,        // row
+        index / rows         // col
+    };
+  }
+}
+
+static f32 stagger_grid_delay(i32 index, i32 count,
+                              const StaggerOptions &stagger_opts)
+{
+  auto [row, col] =
+      grid_position(index, stagger_opts.grid.rows, stagger_opts.grid.cols,
+                    stagger_opts.grid.direction);
+
+  if (!stagger_opts.easing)
+  {
+    return stagger_opts.easing(row, col, index, count);
+  }
+  else
+  {
+    // Default grid-based delay calculation
+    f32 normalized_row = static_cast<f32>(row) / (stagger_opts.grid.rows - 1);
+    f32 normalized_col = static_cast<f32>(col) / (stagger_opts.grid.cols - 1);
+
+    switch (stagger_opts.grid.direction)
+    {
+      case GridDirection::RowMajor:
+        return normalized_row + (normalized_col * 0.5f);
+      case GridDirection::ColumnMajor:
+        return normalized_col + (normalized_row * 0.5f);
+    }
+  }
+}
+
+static f32
+    stagger_linear_delay(i32 i, i32 count,
+                         const StaggerOptions &stagger_opts = StaggerOptions{})
+{
+  return stagger_opts.easing ? stagger_opts.easing(0, 0, i, count) :
+                               static_cast<f32>(i) / count;
+}
+
+static f32 stagger_delay(i32 i, i32 count,
+                         const StaggerOptions &stagger_opts = StaggerOptions{})
+{
+  bool is_grid_based = stagger_opts.grid.rows > 1 || stagger_opts.grid.cols > 1;
+
+  if (is_grid_based)
+  {
+    return stagger_grid_delay(i, count, stagger_opts);
+  }
+
+  return stagger_linear_delay(i, count, stagger_opts);
+}
+
+enum class SegmentType
+{
+  Basic,            // Single Animation
+  Keyframed,        // Multiple Keyframes
+  Parallel,         // Multiple parallel animations
+  Staggered         // Staggered animations
+};
+
+template <Animatable T>
+struct TimelineSegment
+{
+  std::string label    = "default";
+  f32         start    = 0.0f;
+  f32         duration = 0.0f;
+  SegmentType type     = SegmentType::Basic;
+
+  virtual ~TimelineSegment()                      = default;
+  virtual void        update(f32 time, f32 delta) = 0;
+  virtual void        reset()                     = 0;
+  virtual std::string get_label()
+  {
+    return label;
+  }
+  virtual T value() const = 0;
+
+private:
+  T interpolated_value;
+};
+
+template <Animatable T>
+struct BasicSegment : TimelineSegment<T>
+{
+  Animation<T> animation;
+
+  BasicSegment(const T &start, const T &end, AnimationConfig config) :
+      animation(start, end, config, this->onUpdate)
+  {
+    this->duration = config.duration;
```

Comment:
this should be part of the element constructors

Reply:
`this->duration` is part of the `TimelineSegment` struct, it's not a direct member of `BasicSegment`